### DISCUSSION TITLE
Are you sure that wheels for apple silicon works?

### DISCUSSION THREAD

**Discussion by MarshalX** [pypa/cibuildwheel#647] [Q&A] [answered]
I think that delocate doesnt support arm architecture. To build wheels I used GitHub Actions with `CIBW_ARCHS_MACOS: "x86_64 universal2 arm64"`. Only wheels for x86_64 works on **intel** macbooks. I tried to run universal and arm wheels under native python 3.9 on apple silicon and under rosetta terminal too. Doesnt work 

![obraz](https://user-images.githubusercontent.com/15520314/115700816-610af880-a367-11eb-893e-8046d1e7839c.png)

![obraz](https://user-images.githubusercontent.com/15520314/115700823-62d4bc00-a367-11eb-8dd7-b6b08c6e6dfc.png)

**Comment by Czaki** [comment, marked as answer]
As You wrote in the previous topic you have part of the code compiled by CMake. 

As I see in your workflow file you compile multiple libraries before building the wheel. 
Are you sure that all these libraries are compiled to universal binaries? By default, they are build to be native ones. 

`delocate` collect dependent libraries but Your log suggests that you compiled everything into one binary so `delocate` has nothing to do in your case.

**Reply by MarshalX** [reply]
I agree with you. Building dependencies not for the arm is very strange. And I was very surprised when I was able to run the library locally without any problems. But it starts without a wheel. I build it like this: `python setup.py build`. I'm using [bootstrap file](https://github.com/MarshalX/tgcalls/blob/main/pytgcalls/tgcalls_dev.py) to import a .so file as a python module and its works.

**Reply by Czaki** [reply]
If you build wheel on apple arm machine it will work, because all libraries are build native. 

MacOS poorly report library load error. You should inspect this files manually. 

If I good remember python build system does not fully support building wheels using CMake. You still need to be sure if all cmake flags are set properly. I think that you need to ensure it in your `setup.py` file (check target architecture an then set proper flags.

As example of package which uses cibuildwheel and has external dependencies is python-snappy https://github.com/andrix/python-snappy 

You may check it for inspiration. 

To expect a properly working cibuildwheel you need to ensure that Your build system fully works locally. 
So when you have an intel and arm machine you need to build a universal wheel locally on an intel machine and check it on arm one. Then if everything works properly you may try to setup build process using `cibuildwheel`

**Reply by MarshalX** [reply]
I researched the topic and was able to correct build the wheel locally. Unfortunately, the dependencies cannot be builded on the GitHub CI for the ARM... I need an emulator like QEMU for Linux. I will upload from the local machine. Thanks for the answer!

https://github.com/MarshalX/tgcalls/commit/445471a7765ddc897e9da877131823776b296c5f#commitcomment-49858327

**Reply by MarshalX** [reply]
btw https://cmake.org/cmake/help/latest/variable/CMAKE_APPLE_SILICON_PROCESSOR.html

**Reply by henryiii** [reply]
Isn't that the other direction? That looks like how you build x86 code _from_ AS. https://github.com/joerick/cibuildwheel/issues/582 has useful info for AS.

**Reply by henryiii** [reply]
Some but not all packages can be cross-compiled to AS from Intel if you set the right settings (I think linked to from that issue).
